src/rules/no-assign-ref.ts

```diff
@@ -0,0 +1,38 @@
+import { createRule } from '../utils/create-rule';
+
+type MessageIds = 'noAssignRef';
+type Options = [];
+
+export const noAssignRef = createRule<MessageIds, Options>({
+    meta: {
+        type: 'suggestion',
+        fixable: 'code',
+        docs: {
+            description: 'Disallow assigning by reference',
+        },
+        messages: {
+            noAssignRef:
+                "Assigning by reference is not allowed.",
+        },
+        schema: [],
+    },
+
+    create(context) {
+        return {
+            "assignref"(node) {
```

Comment:
support all these cases:
https://astexplorer.net/#/gist/6042c42cc4fe960fb8a6bc4c6e70fdb3/17d7ab7ea4ea547ebc50264175023b40c730b9ac

(maybe in the future we can add options to configure which types to disallow, but for now let's disallow anything)